decode/data.ts

```diff
@@ -1,23 +1,76 @@
-import { Event, IDecodedEvent, IPageData, IPingData, ITagData, Token } from "../types/data";
+import { Envelope, Event, MetricData, PageData, PingData, State, SummaryData, TagData, Token, Upload } from "../types/data";
+import { DataEvent, SummaryEvent } from "../types/decode";
 
-export default function(tokens: Token[]): IDecodedEvent {
+let summaries: { [key: number]: SummaryData[] } = null;
+const SUMMARY_THRESHOLD = 250;
+
+export function reset(): void {
+    summaries = {};
+}
+
+export function decode(tokens: Token[]): DataEvent {
     let time = tokens[0] as number;
     let event = tokens[1] as Event;
     switch (event) {
         case Event.Page:
-            let page: IPageData = {
+            let page: PageData = {
                 timestamp: tokens[2] as number,
                 elapsed: tokens[3] as number,
                 url: tokens[4] as string,
-                title: tokens[5] as string,
-                referrer: tokens[6] as string
+                referrer: tokens[5] as string,
+                lean: tokens[6] as number,
             };
             return { time, event, data: page };
         case Event.Ping:
-            let ping: IPingData = { gap: tokens[2] as number };
+            let ping: PingData = { gap: tokens[2] as number };
             return { time, event, data: ping };
         case Event.Tag:
-            let tag: ITagData = { key: tokens[2] as string, value: tokens[3] as string };
+            let tag: TagData = { key: tokens[2] as string, value: tokens[3] as string };
             return { time, event, data: tag };
+        case Event.Metric:
+            let i = 0;
+            let metrics: MetricData = {};
+            while (i < tokens.length) {
+                metrics[tokens[i++] as number] = tokens[i++] as number;
+            }
+            return { time, event, data: metrics };
+    }
+}
+
+export function envelope(tokens: Token[]): Envelope {
+    return {
+        elapsed: tokens[0] as number,
+        sequence: tokens[1] as number,
+        version: tokens[2] as string,
+        projectId: tokens[3] as string,
+        userId: tokens[4] as string,
+        sessionId: tokens[5] as string,
+        pageId: tokens[6] as string,
+        upload: tokens[7] as Upload,
+        end: tokens[8] as State
```

Comment:
>State [](start = 26, length = 5)

State is very broad name. Could you indicate what kind of state this represents and prefix the name with that?

Reply:
Thanks and called it BooleanFlag.